### PR TITLE
Add new argument processing style, restructure files.

### PR DESCRIPTION
**Pull request by Varriount**
This pull requests adds new argument processing code that makes addition and modification of modes much easier. Furthermore, modes are now separated by file, and are used by the main code via methods. I've removed the table in favor of a case statement, and incorporated backwards-compatibility.

### CODE REVIEW

File: nimsuggest.nim [hunk 1]
```diff
@@ -8,358 +8,190 @@
...
+  var parser = initOptParser()
+  result = CmdLineData(
+      mode: "",
+      nimsuggestSwitches: @[],
+      modeSwitches: @[],
+      compilerSwitches: @[],
+      projectPath: "",
+    )
+
+  result.mode = "stdin"
```
Comment (Araq) [inline nimsuggest.nim]: Old command line interface doesn't default to 'stdin'.

Comment (Varriount) [inline nimsuggest.nim]: Uh, yes it does? At least, it does for me.

**Comment by Varriount** [comment]
@yuutayamada Fixed.

**Comment by yuutayamada** [comment]
@Varriount I added missing function from original nimsuggest's serveEpc function because
nimsuggest didn't start correctly from emacs.

Also could you check following command? I got a strange error...

```sh
-[16410]% ./nimsuggest --nim.define:release stdin sexp.nim
SIGSEGV: Illegal storage access. (Attempt to read from nil?)
```

or am I specifying wrong format?
(I would use --nim.define:nimscript and --nim.define:nimconfig for auto-complete feature
in nimscript files.)

**Comment by Varriount** [comment]
@yuutayamada Fixed. The compiler initialization logic was being later than needed.

**Comment by yuutayamada** [comment]
@Varriount Thank you for quick fix. I made sure auto-completion feature in a nimscript file
and in Emacs. It worked fine after I changed nimsuggest options to new format.

**Comment by yuutayamada** [comment]
@Varriount I think `oldProcessCmdLine` function doesn't treat well options if 
users only specify --help, -h, and --version options.

```nim
  if normalize(cmdLineData.mode) notin ["tcp", "epc", "stdin"]:
    cmdLineData = oldProcessCmdLine()
```

```sh
-[16459]% ./nimsuggest -h
Error: Project path not supplied
-[16460]% ./nimsuggest --help
Error: Project path not supplied
-[16461]% ./nimsuggest --version
Error: Project path not supplied
```
(should treat --help, --h, and --version differently?)

**Comment by Varriount** [comment]
@yuutayamada Done.

**Comment by yuutayamada** [comment]
@Varriount Thanks. It worked fine 
(I wanted --version to distinguish old nimsuggest and new one)

**Comment by yuutayamada** [comment]
@Varriount I found a regression when I tried `chk` option

(below checking has no error on master branch)
```sh
-[16384]% ./nimsuggest stdin nimsuggest.nim
suggest Stdin Mode                                                                  
Project file: "nimsuggest.nim"
Usage: sug|con|def|use|dus|chk|highlight|outline file.nim[;dirtyfile.nim]:line:col
Type 'quit' to quit, 'debug' to toggle debug mode on/off, and 'terse'
to toggle terse mode on/off.

> chk nimsuggest.nim
Hint: nimsuggest [Processing]
nimsuggest.nim(84, 39) Error: type mismatch: got (nimsuggest.CmdLineData)
but expected one of: 
stdinMode.initStdinModeData(cmdline: nimsuggest.CmdLineData)

nimsuggest.nim(86, 35) Error: type mismatch: got (nimsuggest.CmdLineData)
but expected one of: 
tcpMode.initTcpModeData(cmdline: nimsuggest.CmdLineData)

nimsuggest.nim(88, 35) Error: type mismatch: got (nimsuggest.CmdLineData)
but expected one of: 
epcMode.initEpcModeData(cmdline: nimsuggest.CmdLineData)
```

**Comment by Varriount** [comment]
@yuutayamada Are you sure that you are running the chk command from both binaries on the same version of nimsuggest?
I have two versions of nimsuggest, nimsuggest-old.exe and nimsuggest-new.exe . Running the chk command from either of them on a version of nimsuggest.nim from the master branch yields no error, while running them on a version of nimsuggest.nim from this PR yields errors from both.

**Comment by yuutayamada** [comment]
@Varriount I made sure once again because I couldn't remember the situation precisely...

I newly cloned nimsuggest to `~/local/vcs/github.com/nim-lang/nimsuggest` then:
```sh
§ uta@Penguin §☁|origin/major-restructure § 2016/03/21 21:24 §
§~/local/vcs/github.com/nim-lang/nimsuggest §
λ nim --version
Nim Compiler Version 0.13.1 (2016-03-20) [Linux: amd64]
Copyright (c) 2006-2016 by Andreas Rumpf

git hash: 4457b0980fc974efae25c539bad98b9b971299db
active boot switches: -d:release

§ uta@Penguin §☁|origin/major-restructure § 2016/03/21 21:24 §
§~/local/vcs/github.com/nim-lang/nimsuggest §
λ ls ..
Nim  c2nim  nimble  nimsuggest

§ uta@Penguin §☁|origin/major-restructure § 2016/03/21 21:24 §
§~/local/vcs/github.com/nim-lang/nimsuggest §
λ nim e compile_without_nimble.nims

... compiling ...
(I changed ../nim to ../Nim for my directory structure in that .nims file) 

§ uta@Penguin §☁|origin/major-restructure § 2016/03/21 21:30 §
§~/local/vcs/github.com/nim-lang/nimsuggest §
λ .nimsuggest stdin nimsuggest.nim
...
chk nimsuggest.nim
```
then I got the same error. Before I check,  I killed nimsuggest process for Emacs and
I'm managing the repository in different place. (I'm using absolute path)

**Comment by yuutayamada** [comment]
@Varriount I merged that EPC branch here as I said on irc
https://github.com/yuutayamada/nimsuggest/tree/major-restructure
(Should I push here?)

**Comment by yuutayamada** [comment]
I'm not quite sure, but if I specified nimsuggest.nim.cfg first, 
it worked correctly...

```sh
-[16787]% ./nimsuggest stdin nimsuggest.nim.cfg
Running Nimsuggest Stdin Mode
Project file: "nimsuggest.nim.cfg"
Usage: sug|con|def|use|dus|chk|highlight|outline file.nim[;dirtyfile.nim]:line:col
Type 'quit' to quit, 'debug' to toggle debug mode on/off, and 'terse'
to toggle terse mode on/off.

> chk /home/uta/local/vcs/github.com/nim-lang/nimsuggest
Hint: nimsuggest [Processing]
[1]    3762 bus error  ./nimsuggest stdin nimsuggest.nim.cfg
-[16788]% 
```

**Comment by yuutayamada** [comment]
Sorry, never mind. In above check, I didn't input file name (nimsuggest.nim) and
I was blind that bus error message.

**Comment by Varriount** [comment]
@yuutayamada Look at [this](https://gist.github.com/Varriount/b1498c167b5ba19ac8d6)

**Comment by yuutayamada** [comment]
@Varriount What do you think about that result? I don't blight to the nim compiler,
so I'm not sure whether it's a compiler bug or something that fixable in nimsuggest 
repository. (I checked main.nim in nim repo to see difference between `check` command and
`chk`, but I couldn't figure out... both are using similar logic for me)

**Comment by yuutayamada** [comment]
@Varriount old style command line parser overrides --epc option to standin mode.

```sh
./nimsuggest --epc nimsuggest.nim
(starts stdin mode)
```

(Sorry delayed conforming, I noticed command line failure earlier but I thought it was because of missing --nim. prefix in my Emacs)

**Comment by Araq** [comment]
I don't accept restructuring PRs as long as we don't have tests.

**Comment by Araq** [comment]
I don't accept restructuring PRs as long as we don't have tests.

**Comment by Araq** [comment]
I really think it's overkill.

**Pull request closed without merging**